### PR TITLE
Add basic context parallelism: 1/3

### PR DESCRIPTION
**Pull request by liurupeng**
1. Add reorder_sequence function to rearrange q/k/v sequences and reorder it back
2. Add Loadbalanced casual mask 
3. Allow passing mask in the splash attention kernel for torchax
4. Add tests for these functions

Next PR will be adding sharding support for context parallelism

Below is a brief write up of how this supports CP:
You could read some context here: https://insujang.github.io/2024-09-20/introducing-context-parallelism/

WechatIMG178

### CODE REVIEW

File: torchprime/experimental/torchax_models/utils/parallelism_utils.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import jax.numpy as jnp
+import numpy as np
+from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
+
+
+@partial(jax.jit, static_argnames=("cp_size", "seq_dim", "to_contiguous"))
+def reorder_sequence(
+  tensor, cp_size: int, seq_dim: int = 1, to_contiguous: bool = False
+):
+  """Reorders the sequence of the tensor. For example, with cp_size=2, and to_contiguous
```
[the remark this answers is not in the document]
Comment (vlasenkoalexey) [reply torchprime/experimental/torchax_models/utils/parallelism_utils.py:13]: +1

File: torchprime/experimental/torchax_models/utils/parallelism_utils.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
+from functools import partial
```
[the remark this answers is not in the document]
Comment (vlasenkoalexey) [reply torchprime/experimental/torchax_models/utils/parallelism_utils.py:1]: I see file under utils below, can this one be removed?

**Comment by vlasenkoalexey** [comment]
I'd appreciate some short writeup on how this sequence re-shuffling and updated mask would be used for context parallelism.

File: torchprime/experimental/torchax_models/utils/parallelism_utils.py (hunk 2 above)
Comment (tengyifei) [inline torchprime/experimental/torchax_models/utils/parallelism_utils.py:1]: In relation to folks' comment above, my suggestion is to write a single `parallelism_utils` using vanilla PyTorch. Both PyTorch/XLA and torchax can run vanilla PyTorch, just with different lowering.

If you have to write some JAX code, you may use `torch_xla.core.xla_builder.call_jax`. `call_jax` supports both PyTorch/XLA and torchax tensors. So both frameworks can run this function.

**Review by tengyifei — CHANGES_REQUESTED** [review]

File: torchprime/experimental/torchax_models/utils/parallelism_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liurupeng) [reply torchprime/experimental/torchax_models/utils/parallelism_utils.py:1]: the two files is slightly different, they use different jax/np functions

Comment (liurupeng) [reply torchprime/experimental/torchax_models/utils/parallelism_utils.py:1]: thanks, let me try

**Review by jialei777 — APPROVED** [review]

**Review by tengyifei — COMMENTED** [review]
LGTM
